src/Tools/dotnet-monitor/CollectionRules/CollectionRuleService.cs

```diff
@@ -55,42 +72,109 @@ public async ValueTask DisposeAsync()
                 throw new ArgumentNullException(nameof(endpointInfo));
             }
 
+            IProcessInfo processInfo = await ProcessInfoImpl.FromEndpointInfoAsync(endpointInfo);
+
             IReadOnlyCollection<string> ruleNames = _provider.GetCollectionRuleNames();
 
+            CollectionRuleContainer container = new(
+                _serviceProvider,
+                _logger,
+                processInfo);
+
+            lock (_containers)
+            {
+                _containers.Add(container);
+            }
+
             if (ruleNames.Count > 0)
             {
-                KeyValueLogScope logScope = new();
-                logScope.AddCollectionRuleEndpointInfo(endpointInfo);
-                // Constrain the scope of the log scope to just the log call so that the log scope
-                // is not captured by the rule execution method.
-                using (_logger.BeginScope(logScope))
+                await container.StartRulesAsync(ruleNames, token);
+            }
+        }
+
+        private async Task ReapplyRulesAsync(CancellationToken token)
+        {
+            bool rulesChanged = false;
+            while (!token.IsCancellationRequested)
+            {
+                if (!rulesChanged)
                 {
-                    _logger.ApplyingCollectionRules();
+                    using EventTaskSource<EventHandler> rulesChangedTaskSource = new(
+                        callback => (s, e) => callback(),
+                        handler => _provider.RulesChanged += handler,
+                        handler => _provider.RulesChanged -= handler,
+                        token);
+
+                    // Wait for the rules to be changed
+                    await rulesChangedTaskSource.Task;
+
+                    rulesChanged = false;
                 }
 
-                IProcessInfo processInfo = await ProcessInfoImpl.FromEndpointInfoAsync(endpointInfo);
+                _logger.CollectionRuleConfigurationChanged();
 
-                CollectionRuleContainer container = new(
-                    _serviceProvider,
-                    _logger,
-                    processInfo);
-                _containers.Add(container);
+                // Get a copy of the container list to avoid having to
```

Comment:
So the scenario here is if we are restarting the existing set of containers, another one may be added correct?

Reply:
Yes, because configuration change can happen async to a process being discovered, I wanted some guards are modifying the collection of containers.